### ISSUE TITLE
use the Mutable Module to avoid bucketing

### ISSUE THREAD

**Issue by lilhope** [dmlc/mxnet#5014]
Hi dmlc groups , now I want to train a LSTM model, for some reasons, I must set the batch size =1, that means I'll only provide one sentence to the network. Now I want to use the MutableModule(see:`/mxnet/example/rcnn/rcnn/core/module`) to avoid `bucketing`,This is my code(simple):  
```
    def sym_gen(seq_len):
          """ some symbols
          """
          return symbol
    #MutableModule, i'll just list the forward function
    class MuatableModule(BaseModule):
             def forward(databatch,istrain=True):
                   seq_len = databatch.bucket_key #bucket_key is the length of the sentence
                   sym = sym_gen(seq_len)
                   module = Module(sym) # create a new module 
                   module.bind() #bind
                   self._curr_module = module
                   self._curr_module.forward() #do the forward
```
My question is , for each training,I must create a new module,bind it ,and do forward, backward,update, will it be extremely slow? And is there an better implement ? I think bucketing is a good idea for create less Module,but it seems don't fit my work, because the shape of the training data is changing all the time(not only the sentence,I feed an image to the net in the same time).   Thanks in advance!

**Comment by Godricly** [comment]
Y is bucketing not good for you? it sounds like you are doing ocr.

**Comment by lilhope** [comment]
@Godricly I think bucketing is create some Module in advance,and during training time,chose one module as the current module,but my input is changing all the time(not only the sentence,but also the images), I take the example in `rcnn/core/module` as reference,it check whether the input shape is changed, if so ,rebind the module, code like this: 
``` class MutableModule(BaseModule):
     def forward(self, data_batch, is_train=None):
        assert self.binded and self.params_initialized

        # get current_shapes
        if self._curr_module.label_shapes is not None:
            current_shapes = dict(self._curr_module.data_shapes + self._curr_module.label_shapes)
        else:
            current_shapes = dict(self._curr_module.data_shapes)

        # get input_shapes
        if data_batch.provide_label is not None:
            input_shapes = dict(data_batch.provide_data + data_batch.provide_label)
        else:
            input_shapes = dict(data_batch.provide_data)

        # decide if shape changed
        shape_changed = False
        for k, v in current_shapes.items():
            if v != input_shapes[k]:
                shape_changed = True

        if shape_changed:
            module = Module(self._symbol, self._data_names, self._label_names,
                            logger=self.logger, context=self._context,
                            work_load_list=self._work_load_list,
                            fixed_param_names=self._fixed_param_names)
            module.bind(data_batch.provide_data, data_batch.provide_label, self._curr_module.for_training,
                        self._curr_module.inputs_need_grad, force_rebind=False,
                        shared_module=self._curr_module)
            self._curr_module = module

        self._curr_module.forward(data_batch, is_train=is_train) 
``` 
As you can see, if input shape changes,it will create a new module,bind it,and take it as the current module,the do the forward.Because the input shape is changing all the time, I don't think I can create some Module in advance.

**Comment by Godricly** [comment]
The bucketing only creates new module when the bucket_key is met first time. It's pretty much the same thing you want. The input shape is not a big issue. It depends on how you defining and parsing the bucket_key. For me, I'm using a string as bucket key,with label length,and lstm length concated. 
``` python
    def sym_gen(bucket_key):
        label_length, lstm_length = bucket_key.split(',')
        lstm_length = int(lstm_length)
        label_length = int(label_length)
```

**Comment by lilhope** [comment]
@Godricly , but I don't want to pad my image. If my implement doesn't work, I'll try your method, thank you very much!

**Comment by Godricly** [comment]
good luck
